File: hibernate-jpa/src/test/groovy/io/micronaut/configuration/hibernate/jpa/datasources/db1/BookRepository.java
package io.micronaut.configuration.hibernate.jpa.datasources.db1;

import io.micronaut.configuration.hibernate.jpa.scope.CurrentSession;
import io.micronaut.context.annotation.Context;
import io.micronaut.spring.tx.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Optional;

@Transactional
@Context
public class BookRepository {

    private final EntityManager entityManager;

    public BookRepository(@CurrentSession EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    public Book save(@NotBlank String isdn, @NotBlank String name) {
        Book book = new Book(isdn, name);
        entityManager.persist(book);
        return book;
    }

    @Transactional(readOnly = true) // <3>
    public Optional<Book> findById(@NotNull Long id) {
        return Optional.ofNullable(entityManager.find(Book.class, id));
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/hibernate/session/BookRepository.kt
package io.micronaut.docs.hibernate.session

// tag::imports[]
import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Singleton
import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
// end::imports[]

// tag::clazz[]
@Singleton
open class BookRepository {

    @PersistenceContext
    lateinit var entityManager: EntityManager // <1>

    @PersistenceContext(name = "other")
    lateinit var otherManager: EntityManager // <2>

    @Transactional // <3>
    open fun save(title: String): Book {
        val book = Book(title)
        entityManager.persist(book)
        return book
    }

    @Transactional("other") // <4>
    open fun saveToOther(title: String): Book {
        val book = Book(title)
        otherManager.persist(book)
        return book
    }

    @Transactional(readOnly = true)
    open fun findById(id: Long): Book? = entityManager.find(Book::class.java, id)

    @Transactional(value = "other", readOnly = true)
    open fun findInOtherById(id: Long): Book? = otherManager.find(Book::class.java, id)
}
// end::clazz[]
